test/local_creds_test.go

```diff
@@ -0,0 +1,193 @@
+	}
+	defer cc.Close()
+
+	c := testpb.NewTestServiceClient(cc)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	if _, err := c.EmptyCall(ctx, &testpb.Empty{}); err != nil {
+		return fmt.Errorf("EmptyCall(_, _) = _, %v; want _, <nil>", err)
+	}
+
+	return nil
+}
+
+func (s) TestLocalhost(t *testing.T) {
+	err := testE2ESucceed("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Failed e2e test for localhost: %v", err)
+	}
+}
+
+func (s) TestUDS(t *testing.T) {
+	addr := fmt.Sprintf("/tmp/grpc_fullstck_test%d", time.Now().UnixNano())
+	err := testE2ESucceed("unix", addr)
+	if err != nil {
+		t.Fatalf("Failed e2e test for UDS: %v", err)
+	}
+}
+
+type connWrapper struct {
+	net.Conn
+	remote net.Addr
+}
+
+func (c connWrapper) RemoteAddr() net.Addr {
+	return c.remote
+}
+
+type lisWrapper struct {
+	net.Listener
+}
+
+func newLisWrapper(l net.Listener) net.Listener {
+	return &lisWrapper{l}
+}
+
+var remoteAddrs = []net.Addr{
+	&net.IPAddr{
+		IP:   net.ParseIP("10.8.9.10"),
+		Zone: "",
+	},
+	&net.IPAddr{
+		IP:   net.ParseIP("10.8.9.11"),
+		Zone: "",
+	},
+}
+
+func (l *lisWrapper) Accept() (net.Conn, error) {
+	c, err := l.Listener.Accept()
+	if err != nil {
+		return nil, err
+	}
+	return connWrapper{c, remoteAddrs[0]}, nil
+}
+
+func dialer(target string, t time.Duration) (net.Conn, error) {
+	c, err := net.DialTimeout("tcp", target, t)
+	if err != nil {
+		return nil, err
+	}
+	return connWrapper{c, remoteAddrs[1]}, nil
+}
+
+func testE2EFail(useLocal bool) error {
+	ss := &stubServer{
+		emptyCall: func(context.Context, *testpb.Empty) (*testpb.Empty, error) {
+			return &testpb.Empty{}, nil
+		},
+	}
+
+	sopts := []grpc.ServerOption{grpc.Creds(local.NewCredentials())}
+	s := grpc.NewServer(sopts...)
+	defer s.Stop()
+
+	testpb.RegisterTestServiceServer(s, ss)
+
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return fmt.Errorf("Failed to create listener: %v", err)
+	}
+
+	go s.Serve(newLisWrapper(lis))
+
+	var cc *grpc.ClientConn
+	if useLocal {
+		cc, err = grpc.Dial(lis.Addr().String(), grpc.WithTransportCredentials(local.NewCredentials()), grpc.WithDialer(dialer))
+	} else {
+		cc, err = grpc.Dial(lis.Addr().String(), grpc.WithInsecure(), grpc.WithDialer(dialer))
+	}
+
+	if err != nil {
+		return fmt.Errorf("Failed to dial server: %v, %v", err, lis.Addr().String())
+	}
+	defer cc.Close()
+
+	c := testpb.NewTestServiceClient(cc)
+	ctx, cancel := context.WithTimeout(context.Background(), time.Second)
+	defer cancel()
+
+	_, err = c.EmptyCall(ctx, &testpb.Empty{})
+	return err
+}
+
+func (s) TestClientFail(t *testing.T) {
+	// Use local creds at client-side which should lead to client-side failure.
+	err := testE2EFail(true /*useLocal*/)
+	if err == nil || !strings.Contains(err.Error(), "local credentials rejected connection to non-local address") {
+		t.Fatalf("testE2EFail(%v) = _; want security handshake fails, %v", false, err)
```

Comment:
This looks wrong.  You have the returned error being printed as the desired error.

Reply:
Good catch. The test is revised.